File: scanner.py
# scanner.py
"""Polymarket slug-based scanner for active 15-minute crypto markets."""

import asyncio
import json
import time
import random
from datetime import datetime, timezone

import aiohttp
from colorama import Fore, Style, init

from config import Config

# Initialize colorama
init(autoreset=True)

GAMMA_MARKET_URL = "https://gamma-api.polymarket.com/markets"

ASSETS = ["btc", "eth", "sol"]   # XRP removed: no Chainlink oracle feed
INTERVAL_SECONDS = 900          # 15 minutes
REFRESH_INTERVAL_SECONDS = 30
RETRY_WAIT_SECONDS = 10
MAX_TIME_REMAINING_MINUTES = 30  # accept current + next upcoming market
SLUG_FETCH_RETRIES = 3          # max per-slug retry attempts
SLUG_FETCH_BASE_DELAY = 5.0     # seconds — doubles on each retry (5s, 10s, 20s)

ACTIVE_MARKETS: dict[str, dict] = {}


# ---------------------------------------------------------------------------
# Public helpers
# ---------------------------------------------------------------------------

def get_active_markets() -> dict[str, dict]:
    """Return the current ACTIVE_MARKETS dict."""
    return ACTIVE_MARKETS


def get_market_count() -> int:
    """Return the number of currently tracked markets."""
    return len(ACTIVE_MARKETS)


def prune_expired_markets() -> None:
    """Remove markets that have been expired for more than 15 seconds."""
    to_remove = [
        cid for cid, mkt in ACTIVE_MARKETS.items()
        if _time_remaining_minutes(mkt.get("end_date", "")) * 60 < -15
    ]
    for cid in to_remove:
        del ACTIVE_MARKETS[cid]
    if to_remove:
        print(f"{Fore.WHITE}[SCANNER] Pruned {len(to_remove)} expired market(s) from tracking.")


# ---------------------------------------------------------------------------
# Slug + timestamp helpers
# ---------------------------------------------------------------------------

def _current_slot_ts() -> int:
    """Return Unix timestamp of the current 15-min slot (floor to boundary)."""
    now_utc = datetime.now(timezone.utc)
    minutes_rounded = (now_utc.minute // 15) * 15
    slot_time = now_utc.replace(minute=minutes_rounded, second=0, microsecond=0)
    return int(slot_time.timestamp())


def _slugs_to_fetch() -> list[tuple[str, str]]:
    """Return (asset, slug) pairs for 4 time offsets × 3 assets = 12 slugs."""
    base = _current_slot_ts()
    offsets = (0, INTERVAL_SECONDS, -INTERVAL_SECONDS, 2 * INTERVAL_SECONDS)
    pairs: list[tuple[str, str]] = []
    for asset in ASSETS:
        for offset in offsets:
            ts = base + offset
            pairs.append((asset, f"{asset}-updown-15m-{ts}"))
    return pairs


# ---------------------------------------------------------------------------
# Parsing helpers
# ---------------------------------------------------------------------------

def _time_remaining_minutes(end_date_str: str) -> float:
    """Return minutes until end_date_str from now (UTC). Negative = expired."""
    try:
        end_dt = datetime.fromisoformat(end_date_str.replace("Z", "+00:00"))
        if end_dt.tzinfo is None:
            end_dt = end_dt.replace(tzinfo=timezone.utc)
        return (end_dt - datetime.now(timezone.utc)).total_seconds() / 60.0
    except (ValueError, AttributeError, TypeError):
        return 0.0


def _parse_json_field(value) -> list:
    """Parse a field that may already be a list or a JSON-encoded string."""
    if isinstance(value, list):
        return value
    if isinstance(value, str):
        try:
            result = json.loads(value)
            return result if isinstance(result, list) else []
        except (json.JSONDecodeError, ValueError):
            return []
    return []


def _parse_market(raw: dict, symbol: str, slug: str) -> dict | None:
    """Parse a Gamma API market dict into our internal schema."""
    if not raw.get("active", False) or raw.get("closed", False):
        return None

    condition_id: str = raw.get("conditionId", raw.get("condition_id", ""))
    if not condition_id:
        return None

    end_date: str = raw.get("endDate", raw.get("end_date_iso", raw.get("endDateIso", "")))
    time_remaining = _time_remaining_minutes(end_date)
    time_remaining_seconds = time_remaining * 60

    if time_remaining > MAX_TIME_REMAINING_MINUTES:
        return None
    if time_remaining_seconds < -15:
        return None

    clob_ids = _parse_json_field(raw.get("clobTokenIds", []))
    if len(clob_ids) < 2:
        return None
    up_token_id = str(clob_ids[0])
    down_token_id = str(clob_ids[1])

    outcome_prices = _parse_json_field(raw.get("outcomePrices", []))
    try:
        up_price = float(outcome_prices[0]) if len(outcome_prices) > 0 else 0.0
        down_price = float(outcome_prices[1]) if len(outcome_prices) > 1 else 0.0
    except (ValueError, TypeError):
        up_price = down_price = 0.0

    try:
        volume = float(raw.get("volume", 0.0))
    except (ValueError, TypeError):
        volume = 0.0

    return {
        "condition_id": condition_id,
        "question": raw.get("question", ""),
        "symbol": symbol,
        "up_token_id": up_token_id,
        "down_token_id": down_token_id,
        "up_price": up_price,
        "down_price": down_price,
        "time_remaining_minutes": round(time_remaining, 2),
        "time_remaining_seconds": round(time_remaining_seconds, 1),
        "end_date": end_date,
        "volume": volume,
        "slug": slug,
        "price_to_beat": 0.0,
    }


# ---------------------------------------------------------------------------
# Fetch helpers
# ---------------------------------------------------------------------------

async def _fetch_slug(session: aiohttp.ClientSession, slug: str) -> dict | None:
    """GET /markets?slug={slug} with exponential backoff on transient errors."""
    delay = SLUG_FETCH_BASE_DELAY
    for attempt in range(1, SLUG_FETCH_RETRIES + 1):
        try:
            async with session.get(
                GAMMA_MARKET_URL,
                params={"slug": slug},
                timeout=aiohttp.ClientTimeout(total=10),
            ) as resp:
                if resp.status == 429:  # rate-limited
                    retry_after = float(resp.headers.get("Retry-After", delay))
                    print(
                        f"{Fore.YELLOW}[SCANNER] Rate-limited on {slug} "
                        f"— waiting {retry_after:.1f}s (attempt {attempt}/{SLUG_FETCH_RETRIES})"
                    )
                    await asyncio.sleep(retry_after)
                    delay *= 2
                    continue
                if resp.status >= 500:  # server error — backoff & retry
                    print(
                        f"{Fore.YELLOW}[SCANNER] HTTP {resp.status} on {slug} "
                        f"— retrying in {delay:.1f}s (attempt {attempt}/{SLUG_FETCH_RETRIES})"
                    )
                    await asyncio.sleep(delay)
                    delay *= 2
                    continue
                resp.raise_for_status()
                data = await resp.json()
            if isinstance(data, list) and data:
                return data[0]
            return None
        except aiohttp.ClientResponseError as exc:
            print(
                f"{Fore.RED}[SCANNER] HTTP error {exc.status} on {slug}: {exc.message} "
                f"— retrying in {delay:.1f}s (attempt {attempt}/{SLUG_FETCH_RETRIES})"
            )
            await asyncio.sleep(delay)
            delay *= 2
        except (aiohttp.ClientConnectorError, aiohttp.ServerDisconnectedError) as exc:
            print(
                f"{Fore.RED}[SCANNER] Connection error on {slug}: {exc} "
                f"— retrying in {delay:.1f}s (attempt {attempt}/{SLUG_FETCH_RETRIES})"
            )
            await asyncio.sleep(delay)
            delay *= 2
        except asyncio.TimeoutError:
            print(
                f"{Fore.RED}[SCANNER] Timeout fetching {slug} "
                f"— retrying in {delay:.1f}s (attempt {attempt}/{SLUG_FETCH_RETRIES})"
            )
            await asyncio.sleep(delay)
            delay *= 2
    print(f"{Fore.RED}[SCANNER] FAILED all {SLUG_FETCH_RETRIES} attempts for {slug} — skipping.")
    return None


# ---------------------------------------------------------------------------
# Refresh loop
# ---------------------------------------------------------------------------

async def _refresh_once(session: aiohttp.ClientSession) -> None:
    """Fetch all 12 slugs, parse, rebuild ACTIVE_MARKETS."""
    global ACTIVE_MARKETS

    slugs = _slugs_to_fetch()
    updated: dict[str, dict] = {}

    for asset, slug in slugs:
        raw = await _fetch_slug(session, slug)  # retries + backoff handled inside
        if raw is None:
            wait_time = random.uniform(5, 10)
            print(f"{Fore.YELLOW}[SCANNER] Backing off {wait_time:.1f}s due to failure on {slug}...")
            await asyncio.sleep(wait_time)
            continue

        parsed = _parse_market(raw, asset, slug)

        if parsed:
            ACTIVE_MARKETS[parsed["condition_id"]] = parsed
            updated[parsed["condition_id"]] = parsed
            print(
                f"{Fore.GREEN}[SCANNER] {slug} "
                f"→ {parsed['time_remaining_minutes']:.1f} min "
                f"({parsed['time_remaining_seconds']:.0f}s) remaining  ACCEPTED"
            )

        # Baseline interval of 3 to 5 seconds between successful scan requests
        await asyncio.sleep(random.uniform(3, 5))

    # Remove any markets that are no longer active in this pass
    for cid in list(ACTIVE_MARKETS.keys()):
        if cid not in updated:
            ACTIVE_MARKETS.pop(cid, None)

    if ACTIVE_MARKETS:
        print(
            f"{Fore.GREEN}{Style.BRIGHT}[SCANNER] "
            f"{len(ACTIVE_MARKETS)} active market(s) tracked."
        )
    else:
        print(f"{Fore.YELLOW}[SCANNER] No active 15-min markets found this cycle.")


async def run_scanner() -> None:
    """Continuously refresh ACTIVE_MARKETS every 60 seconds."""
    async with aiohttp.ClientSession() as session:
        while True:
            try:
                await _refresh_once(session)
                prune_expired_markets()
            except Exception as exc:  # pylint: disable=broad-except
                print(f"{Fore.RED}[SCANNER] Unexpected error: {exc} — retrying in {RETRY_WAIT_SECONDS}s...")
                await asyncio.sleep(RETRY_WAIT_SECONDS)
                continue

            await asyncio.sleep(REFRESH_INTERVAL_SECONDS)


if __name__ == "__main__":
    Config.summary()
    asyncio.run(run_scanner())

File: src/scanner.rs
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::RwLock;
use serde::Deserialize;
use chrono::{DateTime, Utc};
use crate::types::Market;

pub type ActiveMarkets = Arc<RwLock<HashMap<String, Market>>>;

const ASSETS: &[&str] = &["btc", "eth", "sol"];
const INTERVAL_SECONDS: i64 = 900; // 15 minutes
const MAX_TIME_REMAINING_SECONDS: f64 = 5400.0; // 90 minutes (matches Python MAX_TIME_REMAINING_MINUTES=90)

#[derive(Clone)]
pub struct Scanner {
    pub markets: ActiveMarkets,
}

#[derive(Debug, Deserialize)]
pub struct GammaMarket {
    #[serde(rename = "conditionId", default)]
    pub condition_id: String,
    #[serde(default)]
    pub question: String,
    #[serde(default)]
    pub active: bool,
    #[serde(default)]
    pub closed: bool,
    #[serde(rename = "clobTokenIds", default)]
    pub clob_token_ids: String,
    #[serde(rename = "outcomePrices", default)]
    pub outcome_prices: String,
    // Full RFC3339 datetime e.g. "2026-06-09T13:15:00Z"
    #[serde(rename = "endDate", default)]
    pub end_date: String,
}

impl Scanner {
    pub fn new() -> Self {
        Self {
            markets: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    pub async fn get_active_markets(&self) -> HashMap<String, Market> {
        self.markets.read().await.clone()
    }

    pub async fn prune_expired_markets(&self) {
        let mut r = self.markets.write().await;
        let now = Utc::now();
        r.retain(|_, m| {
            if let Some(ref ed_str) = m.end_date {
                if let Ok(ed) = DateTime::parse_from_rfc3339(ed_str) {
                    // Keep markets that expired less than 15 seconds ago (Python: -15s grace)
                    return (ed.with_timezone(&Utc) - now).num_seconds() > -15;
                }
            }
            true
        });
    }

}

/// Compute the Unix timestamp of the current 15-minute slot boundary (floor).
fn current_slot_ts() -> i64 {
    let now = Utc::now();
    let total_secs = now.timestamp();
    (total_secs / INTERVAL_SECONDS) * INTERVAL_SECONDS
}

/// Generate all slug candidates to try for this cycle (matches Python _slugs_to_fetch).
/// Format: "{asset}-updown-15m-{unix_ts}"
fn slugs_to_fetch() -> Vec<(String, String)> {
    let base = current_slot_ts();
    let offsets: [i64; 4] = [0, INTERVAL_SECONDS, -INTERVAL_SECONDS, 2 * INTERVAL_SECONDS];
    let mut pairs = Vec::new();
    for &asset in ASSETS {
        for &offset in &offsets {
            let ts = base + offset;
            pairs.push((asset.to_string(), format!("{}-updown-15m-{}", asset, ts)));
        }
    }
    pairs
}

/// Fetch a single market by its slug from the Gamma API.
async fn fetch_by_slug(client: &reqwest::Client, slug: &str) -> Result<Option<GammaMarket>, ()> {
    let url = format!("https://gamma-api.polymarket.com/markets?slug={}", slug);
    let mut backoff = 1;

    for attempt in 1..=3 {
        match client.get(&url).send().await {
            Ok(res) => {
                let status = res.status();
                if status.is_success() {
                    match res.text().await {
                        Ok(text) => {
                            if text.trim() == "[]" {
                                return Ok(None);
                            }
                            match serde_json::from_str::<Vec<GammaMarket>>(&text) {
                                Ok(mut vec) if !vec.is_empty() => return Ok(Some(vec.remove(0))),
                                Ok(_) => return Ok(None),
                                Err(e) => {
                                    println!("[SCANNER] ❌ JSON Parse error for slug {} on attempt {}: {}", slug, attempt, e);
                                }
                            }
                        }
                        Err(e) => {
                            println!("[SCANNER] ⚠️ Failed to read response body for {}: {}", slug, e);
                        }
                    }
                } else {
                    println!("[SCANNER] ⚠️ HTTP {} for slug {} on attempt {}", status, slug, attempt);
                    if status == reqwest::StatusCode::NOT_FOUND {
                        return Ok(None);
                    }
                }
            }
            Err(e) => {
                println!("[SCANNER] ❌ Network error for slug {} on attempt {}: {}", slug, attempt, e);
            }
        }
        if attempt < 3 {
            tokio::time::sleep(tokio::time::Duration::from_secs(backoff)).await;
            backoff *= 2;
        }
    }
    Err(())
}

pub async fn run_scanner(scanner: Scanner) {
    println!("[SCANNER] Scanner started. Fetching 15-min interval markets...");
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(10))
        .build()
        .unwrap_or_default();

    loop {
        let slugs = slugs_to_fetch();
        let mut new_map = HashMap::new();
        let mut scan_failed = false;
        let now = Utc::now();

        for (asset, slug) in &slugs {
            let raw = match fetch_by_slug(&client, slug).await {
                Ok(Some(m)) => m,
                Ok(None) => continue,
                Err(_) => {
                    scan_failed = true;
                    continue;
                }
            };

            if !raw.active || raw.closed { continue; }
            if raw.condition_id.is_empty() { continue; }

            // Parse end_date and compute time remaining
            let end_dt = match DateTime::parse_from_rfc3339(&raw.end_date) {
                Ok(dt) => dt.with_timezone(&Utc),
                Err(_) => continue,
            };
            let time_remaining_seconds = (end_dt - now).num_milliseconds() as f64 / 1000.0;

            // Python: reject if > MAX_TIME_REMAINING_MINUTES (90min) or expired > 15s
            if time_remaining_seconds > MAX_TIME_REMAINING_SECONDS { continue; }
            if time_remaining_seconds < -15.0 { continue; }

            // Parse token IDs from JSON string
            let token_ids: Vec<String> = match serde_json::from_str(&raw.clob_token_ids) {
                Ok(v) => v,
                Err(_) => continue,
            };
            if token_ids.len() < 2 { continue; }

            // Parse outcome prices
            let prices: Vec<String> = serde_json::from_str(&raw.outcome_prices).unwrap_or_default();
            let up_p: f64 = prices.first().and_then(|s| s.parse().ok()).unwrap_or(0.0);
            let down_p: f64 = prices.get(1).and_then(|s| s.parse().ok()).unwrap_or(0.0);

            let symbol = match asset.as_str() {
                "btc" => "BTC",
                "eth" => "ETH",
                "sol" => "SOL",
                _ => continue,
            };

            // println!("[SCANNER] ✅ {} → {} | {:.1}min remaining | ACCEPTED",
            //     slug, raw.question, time_remaining_seconds / 60.0);

            let m = Market {
                condition_id: raw.condition_id.clone(),
                question: raw.question,
                symbol: symbol.to_string(),
                slug: slug.clone(),
                up_token_id: token_ids[0].clone(),
                down_token_id: token_ids[1].clone(),
                up_price: up_p,
                down_price: down_p,
                end_date: Some(end_dt.to_rfc3339()),
                time_remaining_seconds,
            };
            new_map.insert(m.condition_id.clone(), m);

            // Avoid hammering the API — small delay between slug fetches
            tokio::time::sleep(tokio::time::Duration::from_millis(500)).await;
        }


        {
            let mut r = scanner.markets.write().await;
            if !scan_failed {
                *r = new_map;
            } else {
                for (k, v) in new_map {
                    r.insert(k, v);
                }
            }
        }
        scanner.prune_expired_markets().await;

        /*
        if count > 0 {
            println!("[SCANNER] {} active 15-min market(s) tracked.", count);
        } else {
            println!("[SCANNER] ⚠️ No active 15-min markets found this cycle.");
        }
        */

        tokio::time::sleep(tokio::time::Duration::from_secs(30)).await;
    }
}
